### DISCUSSION TITLE
Quickly change a body to non-solid from a Body pointer; possible?

### DISCUSSION THREAD

**Discussion by XephStream** [jrouwe/JoltPhysics#787] [Q&A] [answered]
Hey all, I'm switching from PhysX to Jolt in my game engine, and with my current system I've got the ability for coders to change a rigidbody's solidity to NONE, TRIGGER, or NORMAL (solid).

With PhysX I achieve this as such:

```
// inside SetSolidType()...
// ...
// set the properties on every shape of this body
for (unsigned i = 0; i < numActorShapes && i < Physics::MaxRigidBodyShapes; i++)
{
	switch (m_Solid)
	{
	default: break;
	case Physics::SOLID_NONE:
		shapes[i]->setFlag(PxShapeFlag::eSCENE_QUERY_SHAPE, false); // don't show up in queries
		shapes[i]->setFlag(PxShapeFlag::eSIMULATION_SHAPE, false);  // don't collide with anything
		shapes[i]->setFlag(PxShapeFlag::eTRIGGER_SHAPE, false);     // don't be a trigger volume
		break;

	case Physics::SOLID_TRIGGER:
		shapes[i]->setFlag(PxShapeFlag::eSCENE_QUERY_SHAPE, false); // don't show up in queries
		shapes[i]->setFlag(PxShapeFlag::eSIMULATION_SHAPE, false);  // don't collide with anything
		shapes[i]->setFlag(PxShapeFlag::eTRIGGER_SHAPE, true);      // be a trigger volume
		break;

	case Physics::SOLID_NORMAL:
		shapes[i]->setFlag(PxShapeFlag::eSCENE_QUERY_SHAPE, true);  // show up in queries
		shapes[i]->setFlag(PxShapeFlag::eSIMULATION_SHAPE, true);   // collide with other solid shapes
		shapes[i]->setFlag(PxShapeFlag::eTRIGGER_SHAPE, false);     // don't be a trigger volume
		break;
	}
}
```

However, with a JPH::Body, all I've found so far is `SetIsSensor()`, which is great, but aside from turning that on and off, I can't seem to find the "right" way to make a body still exist and be active, but have no collision response or sensor behavior whatsoever. I've found `SetCollisionGroup` on `BodyInterface`, and of course there are the Layer filtering classes, but I'm hoping there's a way to achieve this quickly using a `Body` pointer directly.

If not, how would you approach this? Layers / collision groups? Work up at the body interface level instead?

**Comment by jrouwe** [comment, marked as answer]
The best way would be to change the collision layer of a body through ```BodyInterface::SetObjectLayer``` with a layer that collides with nothing.

Alternatively you could extend ```GroupFilter``` and implement your own version that uses e.g. 1 bit of ```CollisionGroup::SetGroupID``` or ```CollisionGroup::SetSubGroupID``` to indicate that no collision is desired. Note that object queries like ray casts do not check against the group filter, this is for simulation only. Because the group filter is evaluated later than the object layer, the 2nd method will have more CPU overhead.

**Reply by XephStream** [reply]
Thanks. I've implemented a NONE layer and now use the body interface alongside body pointers, and it works great.

Although, I think the addition of a simple `SetCollisionsEnabled(bool enabled)` to the Body object would come in handy for situations like this :)
